apps/meteor/app/lib/server/functions/setUsername.ts

```diff
@@ -67,78 +68,81 @@ export const setUsernameWithValidation = async (userId: string, username: string
+			return user;
 		}
-	} catch (e: any) {
-		SystemLogger.error(e);
-	}
-	// Set new username*
-	// TODO: use updater for setting the username and handle possible side effects in addUserToRoom
-	await Users.setUsername(user._id, username);
-	user.username = username;
-
-	if (!previousUsername && settings.get('Accounts_SetDefaultAvatar') === true) {
-		const avatarSuggestions = await getAvatarSuggestionForUser(user);
-		let avatarData;
-		let serviceName = 'gravatar';
-
-		for (const service of Object.keys(avatarSuggestions)) {
-			avatarData = avatarSuggestions[service];
-			if (service !== 'gravatar') {
-				serviceName = service;
-				break;
+		const previousUsername = user.username;
+		// Check username availability or if the user already owns a different casing of the name
+		if (!previousUsername || !(username.toLowerCase() === previousUsername.toLowerCase())) {
+			if (!(await checkUsernameAvailability(username))) {
+				return false;
 			}
 		}
+		// If first time setting username, send Enrollment Email
+		try {
+			if (!previousUsername && user.emails && user.emails.length > 0 && settings.get('Accounts_Enrollment_Email')) {
+				setImmediate(() => {
+					Accounts.sendEnrollmentEmail(user._id);
+				});
+			}
+		} catch (e: any) {
+			SystemLogger.error(e);
+		}
+		// Set new username*
+		// TODO: use updater for setting the username and handle possible side effects in addUserToRoom
+		await Users.setUsername(user._id, username, { session });
+		user.username = username;
+
+		if (!previousUsername && settings.get('Accounts_SetDefaultAvatar') === true) {
+			const avatarSuggestions = await getAvatarSuggestionForUser(user);
+			let avatarData;
+			let serviceName = 'gravatar';
+
+			for (const service of Object.keys(avatarSuggestions)) {
+				avatarData = avatarSuggestions[service];
+				if (service !== 'gravatar') {
+					serviceName = service;
+					break;
+				}
+			}
 
-		if (avatarData) {
-			await setUserAvatar(user, avatarData.blob, avatarData.contentType, serviceName, undefined, updater);
+			if (avatarData) {
+				await setUserAvatar(user, avatarData.blob, avatarData.contentType, serviceName, undefined, updater);
+			}
 		}
-	}
 
-	// If it's the first username and the user has an invite Token, then join the invite room
-	if (!previousUsername && user.inviteToken) {
-		const inviteData = await Invites.findOneById(user.inviteToken);
-		if (inviteData?.rid) {
-			await addUserToRoom(inviteData.rid, user);
+		// If it's the first username and the user has an invite Token, then join the invite room
+		if (!previousUsername && user.inviteToken) {
+			const inviteData = await Invites.findOneById(user.inviteToken);
+			if (inviteData?.rid) {
+				await addUserToRoom(inviteData.rid, user);
+			}
 		}
-	}
 
-	void api.broadcast('user.nameChanged', {
-		_id: user._id,
-		name: user.name,
-		username: user.username,
-	});
+		void api.broadcast('user.nameChanged', {
+			_id: user._id,
+			name: user.name,
+			username: user.username,
+		});
```

Comment:
This should happen after transaction commits,otherwise client may get invalid updates.

Reply:
Yep. One of the places with side effects that are not so easy to extract.